bootloader/lkenv/lkenv.go

```diff
@@ -0,0 +1,406 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package lkenv
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"os"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+)
+
+const SNAP_BOOTSELECT_VERSION = 0x00010001
+
+// const SNAP_BOOTSELECT_SIGNATURE ('S' | ('B' << 8) | ('s' << 16) | ('e' << 24))
+const SNAP_BOOTSELECT_SIGNATURE = 0x53 | 0x42<<8 | 0x73<<16 | 0x65<<24
+const SNAP_NAME_MAX_LEN = 256
+const SNAP_MODE_LENGTH = 8
+
+/* number of available boot partitions */
+const SNAP_BOOTIMG_PART_NUM = 2
+
+/* Default boot image file name to be used from kernel snap */
+const BOOTIMG_DEFAULT_NAME = "boot.img"
+
+// for accessing the 	Bootimg_matrix
+const (
+	MATRIX_ROW_PARTITION = 0
+	MATRIX_ROW_KERNEL    = 1
+)
+
+/**
+ * Following structure has to be kept in sync with c structure defined by
+ * include/snappy-boot_v1.h
+ * c headerfile is used by bootloader, this ensures sync of  the environment
+ * between snapd and bootloader
+
+ * when this structure needs to be updated,
+ * new version should be introduced instead together with c header file,
+ * which is to be adopted by bootloader
+ *
+ * !!! Support for old version has to be maintained, as it is not guaranteed
+ * all existing bootloader would adopt new version!
+ */
+type SnapBootSelect_v1 struct {
+	/* Contains value BOOTSELECT_SIGNATURE defined above */
+	Signature uint32
+	/* snappy boot select version */
+	Version uint32
+
+	/* snap_mode, one of: 'empty', "try", "trying" */
+	Snap_mode [SNAP_MODE_LENGTH]byte
+	/* current core snap revision */
+	Snap_core [SNAP_NAME_MAX_LEN]byte
+	/* try core snap revision */
+	Snap_try_core [SNAP_NAME_MAX_LEN]byte
+	/* current kernel snap revision */
+	Snap_kernel [SNAP_NAME_MAX_LEN]byte
+	/* current kernel snap revision */
+	Snap_try_kernel [SNAP_NAME_MAX_LEN]byte
+
+	/* GADGET assets: current gadget assets revision */
```

Comment:
s/GADGET/gadget/ ?

Reply:
yep, I have already fixed this with next update